enterprise/coderd/coderdenttest/proxytest.go

```diff
@@ -47,14 +47,17 @@ func NewWorkspaceProxy(t *testing.T, coderdAPI *coderd.API, owner *codersdk.Clie
 		options = &ProxyOptions{}
 	}
 
-	// HTTP Server
+	// HTTP Server. We have to start this once to get the access URL to start
+	// the workspace proxy with. The workspace proxy has the handler, so the
+	// http server will start with a 503 until the proxy is started.
```

Comment:
A strategy that avoids the mutex is to create a listener, then pass that to `http.Serve()`

```
l := net.Listen("tcp", "127.0.0.1:")
url := "http://" + l.Addr().String()
```

I guess it gets a bit more complex if we need TLS.  Does the wsproxy care whether its TLS?